File: backend/app/schemas/workflow.py
from datetime import datetime
from enum import Enum
from typing import Literal

from pydantic import BaseModel, Field, field_validator


class ICPConfig(BaseModel):
    industry: list[str]
    headcount_min: int = Field(ge=1)
    headcount_max: int = Field(ge=1)
    funding_stages: list[str]
    geography: list[str]
    revenue_min_usd: int | None = Field(default=None, ge=0)
    tech_stack: list[str] | None = None

    @field_validator("headcount_max")
    @classmethod
    def max_greater_than_min(cls, value: int, info):
        minimum = info.data.get("headcount_min")
        if minimum is not None and value < minimum:
            raise ValueError("headcount_max must be greater than or equal to headcount_min")
        return value


class PersonaConfig(BaseModel):
    name: str
    titles: list[str]
    seniority_levels: list[str] | None = None
    priority: int = Field(default=1, ge=1)


class CreateWorkflowRequest(BaseModel):
    name: str
    icp: ICPConfig
    personas: list[PersonaConfig]
    triggers: list[str]


class WorkflowStatus(str, Enum):
    PENDING = "pending"
    RUNNING = "running"
    AWAITING_APPROVAL = "awaiting_approval"
    COMPLETED = "completed"
    FAILED = "failed"


class WorkflowResponse(BaseModel):
    workflow_id: str
    name: str
    status: WorkflowStatus
    results_count: int | None = None
    total_cost_usd: float | None = None
    created_at: datetime
    completed_at: datetime | None = None


class CreateWorkflowResponse(BaseModel):
    workflow_id: str
    status: WorkflowStatus
    websocket_url: str
    estimated_duration_seconds: int = 240


class WorkflowListResponse(BaseModel):
    workflows: list[WorkflowResponse]
    total: int
    page: int
    per_page: int


class ApprovalDecision(BaseModel):
    contact_id: str
    action: Literal["approve", "reject", "edit"]
    reason: str | None = None
    edits: dict | None = None


class SubmitApprovalRequest(BaseModel):
    decisions: list[ApprovalDecision]


class ApprovalResponse(BaseModel):
    workflow_id: str
    status: WorkflowStatus
    processed: int


class WorkflowDetailResponse(BaseModel):
    workflow: dict
    companies: list[dict]
    contacts: list[dict]
    summary_report: dict | list | None = None
    logs: list[dict]

    @field_validator("summary_report", mode="before")
    @classmethod
    def normalise_summary_report(cls, v):
        """Ensure summary_report is always a dict (or None).

        Older workflow runs may have stored the report as a bare list of
        company-summary dicts. Wrap those into the canonical shape that the
        frontend SummaryReport component expects.
        """
        if v is None:
            return None
        if isinstance(v, list):
            return {"companies": v}
        return v

File: frontend/src/types/workflow.ts
import type { Company } from './company';
import type { Contact } from './contact';
import type { LogEntry } from './agent';

export type WorkflowStatus =
  | 'pending'
  | 'running'
  | 'awaiting_approval'
  | 'completed'
  | 'failed';

export interface ICPConfig {
  industry: string[];
  headcount_min: number;
  headcount_max: number;
  funding_stages: string[];
  geography: string[];
  revenue_min_usd?: number;
  tech_stack?: string[];
  min_icp_score?: number;
}

export interface PersonaConfig {
  name: string;
  titles: string[];
  seniority_levels?: string[];
  priority: number;
}

export interface CreateWorkflowRequest {
  name: string;
  icp: ICPConfig;
  personas: PersonaConfig[];
  triggers: string[];
}

export interface CreateWorkflowResponse {
  workflow_id: string;
  status: WorkflowStatus;
  websocket_url: string;
  estimated_duration_seconds: number;
}

export interface CompanySummary {
  company_id: string;
  company_name: string;
  executive_summary: string;
  why_now: string;
  outreach_strategy: Record<string, string>; // persona name -> strategy text
  subject_lines: string[];
  risk_factors: string[];
  recommended_actions: string[];
  confidence_score: number;
}

export interface SummaryReport {
  companies: CompanySummary[];
  generated_at: string;
  total_companies: number;
  total_contacts: number;
}

export interface Workflow {
  workflow_id: string;
  name: string;
  status: WorkflowStatus;
  icp: ICPConfig;
  personas: PersonaConfig[];
  triggers: string[];
  results_count?: number;
  total_cost_usd?: number;
  total_tokens?: number;
  summary_report?: SummaryReport | null;
  created_at: string;
  completed_at?: string;
}

export interface WorkflowListResponse {
  workflows: Workflow[];
  total: number;
  page: number;
  per_page: number;
}

export interface WorkflowDetailResponse {
  workflow: Workflow;
  companies: Company[];
  contacts: Contact[];
  summary_report: SummaryReport | null;
  logs: LogEntry[];
}

export interface ApprovalDecision {
  contact_id: string;
  action: 'approve' | 'reject' | 'edit';
  reason?: string;
  edits?: Partial<Contact>;
}

export interface ApprovalResponse {
  workflow_id: string;
  status: WorkflowStatus;
  processed: number;
}
